crates/red_knot/src/types.rs

```diff
@@ -0,0 +1,91 @@
+#![allow(dead_code)]
+use crate::Name;
+use ruff_index::{newtype_index, IndexVec};
+
+#[newtype_index]
+pub(crate) struct TypeId;
+
+impl TypeId {
+    pub(crate) fn typ(self, env: &TypeEnvironment) -> &Type {
+        env.type_for_id(self)
+    }
+}
+
+/// Arena holding all known types
+#[derive(Default)]
+pub(crate) struct TypeEnvironment {
+    types_by_id: IndexVec<TypeId, Type>,
```

Comment:
I think it will be interesting if we can come up with a more stable id for types than relying on inference order.

Reply:
Yeah, it's a good point, I thought about this. I don't think `TypeId` (as in, the index into this `IndexVec` can be stable; that's not really compatible with lazy type evaluation. But we may want to have another, more stable identifier (based on fully qualified name?) that we use most places instead of `TypeId`.

Reply:
I could also abandon using the `IndexVec` arena in this case? But we will have a lot of types...